ddl/ddl_worker.go

```diff
@@ -123,10 +123,16 @@ func (d *ddl) handleUpdateJobError(t *meta.Meta, job *model.Job, err error) erro
 
 // updateDDLJob updates the DDL job information.
 // Every time we enter another state except final state, we must call this function.
-func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64) error {
+func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64, err error) error {
 	job.LastUpdateTS = int64(updateTS)
-	err := t.UpdateDDLJob(0, job)
-	return errors.Trace(err)
+	updateRawArgs := true
+	// If there is an error when running job and the RawArgs hasn't been decoded wit DecodeArgs,
+	// so we should't replace RawArgs with the marshaling Args.
+	if err != nil && (job.RawArgs != nil && job.Args == nil) {
```

Comment:
why not just use `job.Error != nil`?

Reply:
Because `job.Error` will keep the previous error.